### DISCUSSION TITLE
Multiple `Application` classes in one project?

### DISCUSSION THREAD

**Discussion by tjquinno** [smallrye/smallrye-open-api#946] [Q&A]
(transplanted from the old Google group)

Some time ago on an older release of SmallRye OpenAPI, we noticed trouble when multiple JAX-RS `Aoplication` classes were present. IIRC we would see only one represented in the OpenAPI document produce, and which one could vary from run to run.

I've been doing some things with 2.0.26 recently and it seems as if the scanner and related code are handling multiple apps just fine (see below for a definition of "just fine"). I have seen at least one issue opened about this and closed as not supported (#645). 

Does this in fact work and, if so, is this supported?

I've been working recently with a simple two-application project. I see all endpoints from both apps represented in the OpenAPI document, and the `components` section contains entries for POJOs used from all endpoints.

Certainly the document can contain only one `info` section, even if all apps bear the `@OpenAPIDefinition`, for example. But as stated in #645, the MP OpenAPI spec hints at but doesn't define the "correct" behavior in such cases. So it's developer beware (at least with respect to multiple classes with `@OpenAPIDefinition`) and that seems fine to me.

**Comment by phillip-kruger** [comment]
Hi @tjquinno . 
@MikeEdgar please jump in here, as I do not have all the history.

As far as I know, the spec targets microservices (hence also the `/openapi` url) and thus expects a merged openapi document. The same way that you can also supply a document and merge it with the annotation scanned document. 

In smallrye, this is not really enforced, as the implementations (Quarkus/Wildfly etc) can decide to create multiple separate documents and make them available under a context root. SmallRye allow the scanning of annotations (but the scope / jandex index is supplied by the implementation) and the merging of static documents. So a implementation can basically create a document per application by controlling the jandex index it provides to be scanned.

Not sure if this makes sense. @MikeEdgar @EricWittmann please comment.

**Reply by tjquinno** [reply]
Thanks, @phillip-kruger. Yes, that helps. 

The idea of the implementation controlling the Jandex index passed to SmallRye makes sense for the use case you described. Another intriguing possibility might be to use different scanning include/exclude configuration with the _same_ Jandex index to "focus" SmallRye's attention in multiple invocations, one per application.

But, the use case I have in mind is potentially a bit different. We urge--but do not enforce--app developers to have exactly one JAX-RS `Application` class in their projects. If they have just one, then all is well. 

But suppose a project has two or more `Application` classes in it. I want to make sure there are no problems in SmallRye's OpenAPI implementation. Clearly if multiple `Application` classes have `@OpenAPIDefinition` annotations then the resulting SmallRye model can contain only one `info`, etc. section. 

Other than that, are there any issues in SmallRye itself that would have trouble in this scenario? 

In the limited testing I've done things seem to work fine. I'd be more comfortable, though, if I got a more informed thumbs-up from you all that SmallRye can handle this.

Thanks.

**Reply by phillip-kruger** [reply]
Ok, so everything should work fine. If there are conflicting data in `OpenAPIDefinition` however, I am not sure how the merged document will choose. (@MikeEdgar  ?)

I know that when you provide a static document, if that document contain header info, that would take president over the scanned version. Also, if the implementation or the user provide a filter that set the header info, that will also take president. Lastly, in Quarkus we allow the user to set this using config, and that will override any other (scanning/static/filter). See https://github.com/smallrye/smallrye-open-api/blob/099003f6ef080bbcab459eaf7a7b380f700c93d8/core/src/main/java/io/smallrye/openapi/api/OpenApiConfig.java#L103

Also, there are config that allows you to narrow the classes and packages that should be scanned: https://github.com/smallrye/smallrye-open-api/blob/099003f6ef080bbcab459eaf7a7b380f700c93d8/core/src/main/java/io/smallrye/openapi/api/OpenApiConfig.java#L36

**Reply by tjquinno** [reply]
Exactly. Using the `OpenApiConfig` is what I had in mind when I wrote my earlier comment. I was thinking of multiple invocations of SmallRye, one per `Application`, with the config on each invocation set to _exclude_ all _other_ apps, focusing SmallRye on the app of interest during that particular invocation. Might try that. 

I know the MP OpenAPI spec prescribes which source of information wins (annotation scanning, filter, static document, etc.) if they conflict. 

In the use case I have in mind the conflicts might come from just annotation scanning itself. That's why I wanted to check in here, to be sure SmallRye will not have trouble with that specific situation.

TBH, I think any choice SmallRye might make in resolving such conflicts within annotation processing is probably fine. The developer with multiple `Application` classes all with `@OpenAPIDefinition` annotations should not expect any deterministic result, at least IMO. Just as long as SmallRye doesn't have trouble with this situation.

**Reply by phillip-kruger** [reply]
Ok cool. You are welcome to create a test case that test this, and we can see what happens. We can also scratch in the code to see. And then we can maybe tweak the code to do something else. Also let see what @MikeEdgar says as he knows the implementation better.

**Reply by MikeEdgar** [reply]
At high-level, scanning multiple times with different class/package filters sounds like it will work. Be aware that some top-level items are always scanned (e.g. `@Schema` annotations), so you'll want to also exclude types/beans that may not apply to a particular `Application` (if needed).

**Reply by tjquinno** [reply]
I think I found a problem, at least with one particular use case. We're using 2.0.26.

We have a test project that contains several `Application` classes. Several of them implement `getClasses` and return _the same_ resource class to reflect that they expose the same API but under different paths according to different `@ApplicationPath` settings on each.

`JaxRsAnnotationScanner$scan` invokes `processApplicationClasses` (which gathers annotation information from the app class) and then invokes `processResourceClasses` which processes our two resource classes. I have not stepped through every last line of code, but it looks as if the resource class (I've looked at the paths part of the model in particular) is associated with the most recently-processed application class (saved in the context, maybe??). 

The resulting OpenAPI model contains `paths` entries only for that last application. 

Two quick comments:
1. Having multiple applications in the project might be a bit odd, and the MP OpenAPI spec explicitly washes it hands of this case. Still, it's not impossible that users would do this.
2. Even without multiple applications in the project, the `Application` might implement `getClasses` precisely to limit the resources it wants to expose, even if there are additional ones discoverable in the project. But as written (IIUC)  the model associates ALL resources with the one app, despite the `getClasses` return value.

**Reply by phillip-kruger** [reply]
Yes I don't think we look at the getClasses at all. We just scan all in the classpath. It might be difficult to implement as we would need to get the classes at runtime. @MikeEdgar w.d.y.t ?

**Reply by tjquinno** [reply]
In fact, we did something similar in our code which drives the SmallRye OpenAPI code -- we invoke the application's `getClasses` (and also `getSingletons` to get their classes as well) and drive SmallRye once for each app and use the config to limit the scanning for each invocation of SmallRye.

The problem we ran into is that our code _did not_ also check the resource methods for POJOs used as input or output parameters. So when we built the `include` pattern we accidentally hid those POJOs from the scanning and so the model did not include them. I was hoping there was a clever way to just let SmallRye's scanning do its thing in one invocation, rather than invoking once for each app and having to add yet more logic to handle the POJO parameter situation.

This seems like a tricky problem regardless of whether SmallRye itself would handle it or whether the caller would be responsible.

**Reply by phillip-kruger** [reply]
If we can do it in SmallRye then that would be great. you are welcome to give it a try and do a PR ? The other option is to do this at runtime (like we do the filters).
